cpp_exercise/src/lib/linuxipclib.cc

```diff
@@ -0,0 +1,262 @@
+#include "linuxipclib.h"
+#include "ipcexceptionlib.h"
+
+
+
+/***
+ * Protocol Enum
+ * 
+ * Checking the protocol name as a string
+ * **/
+
+
+ipcprt::Protocol IPC::protocolOptions(const std::string & input){
+    if (input == "pipe") return ipcprt::Protocol::pipe;
+    if (input == "msgqueue") return ipcprt::Protocol::msgqueue;
+    if (input == "shm") return ipcprt::Protocol::shm;
+
+    return ipcprt::Protocol::invalid;
+}
+
+/**
+ * @brief IPCSender Class
+ * */
+
+
+
+/**
+ * @brief IPCSender Pipe Protocol
+ * 
+ * 
+ * 
+ */
+
+
+void IPCSender::ipcpipe(const std::string & filename){
+   
+   try{
+     PipeSender pipes(filename);
+   }catch(IPCException & e){
+     std::cout<<e.what()<<std::endl;
+    exit(EXIT_FAILURE);
+   }
+}
+
+/**
+ * @brief IPCSender MsgQueue Protocol
+ * 
+ * 
+ * 
+ */
+
+
+void IPCSender::ipcmsgqueue(const std::string & filename){
+     
+      MsgQueueSender msgqs(filename);
+
+}
+
+
+/**
+ * @brief IPCSender Constructor for Choosing the one of the protocols
+ * 
+ * 
+ * 
+ */
+
+IPCSender::IPCSender(const std::string & filename, const std::string  & protocol) : _file_name(filename), _protocol(protocol){
+    switch (protocolOptions(_protocol))
+    {
+    case ipcprt::Protocol::pipe:
+        ipcpipe(_file_name);
+        break;
+    
+    case ipcprt::Protocol::msgqueue:
+        ipcmsgqueue(_file_name);
+        break;
+    default:
+        throw IPCException("IPCReceiver ERROR:UNKNOWN IPC PROTOCOL");
+        break;
+    }
+}
+
+
+/**
+ * @brief IPCReceiver Class
+ * 
+ * 
+ * 
+ */
+
+/**
+ * @brief IPCReceiver Pipe Protocol
+ * */
+
+void IPCReceiver::ipcpipe(const std::string & filename){      
+
+   try{
+      PipeReceiver piper(filename);
+   }catch(IPCException & e){
+     std::cout<<e.what()<<std::endl;
+    exit(EXIT_FAILURE);
+   }
+    
+}
+
+/**
+ * @brief IPCReceiver MsgQueue Protocol
+ * */
+
+void IPCReceiver::ipcmsgqueue(const std::string & filename){
+
+  MsgQueueReceiver msgqr(filename);
+
+}
+
+/**
+ * @brief IPCReceiver Choosing Between the Protocols
+ * */
+
+IPCReceiver::IPCReceiver(const std::string & filename, const std::string  & protocol): _file_name(filename), _protocol(protocol){
+     switch (protocolOptions(_protocol))
+    {
+    case ipcprt::Protocol::pipe:
+        ipcpipe(_file_name);
+        break;
+    case ipcprt::Protocol::msgqueue:
+        ipcmsgqueue(_file_name);
+        break;
+    default:
+        throw IPCException("IPCReceiver ERROR:UNKNOWN IPC PROTOCOL");
+        break;
+    }
+
+}
+    
+/**
+ * @brief Command Option Class
+ * 
+ */
+
+ 
+ CommandOption::~CommandOption(){
+
+     _option = 0;
```

Comment:
I don't think setting class attribute in destructor has any sense

Reply:
Fixed